pygsti/forwardsims/torchfwdsim.py

```diff
@@ -0,0 +1,237 @@
+            for superop in superops:
+                superket = superop @ superket
+            circuit_probs = povm_mat @ superket
+            probs.append(circuit_probs)
+        probs = torch.concat(probs)
+        return probs
+    
+    def jac_friendly_circuit_probs(self, *free_params: Tuple[torch.Tensor]):
+        """
+        This function combines parameter reformatting and forward simulation.
+        It's needed so that we can use PyTorch to compute the Jacobian of
+        the map from a model's free parameters to circuit outcome probabilities.
+        """
+        assert len(free_params) == len(self.param_metadata) == self.num_params
+        free_params = {self.param_metadata[i][0] : free_params[i] for i in range(self.num_params)}
+        torch_cache = self.get_torch_cache(free_params, grad=False)
+        probs = self.circuit_probs(torch_cache)
+        return probs
+
+
+class TorchForwardSimulator(ForwardSimulator):
+
+    """
+    A forward simulator that leverages automatic differentiation in PyTorch.
+    """
+    def __init__(self, model : Optional[ExplicitOpModel] = None):
+        if not TORCH_ENABLED:
+            raise RuntimeError('PyTorch could not be imported.')
+        self.model = model
+        super(ForwardSimulator, self).__init__(model)
+
+    @staticmethod
+    def separate_state(model: ExplicitOpModel, layout, grad=False):
+        slm = StatelessModel(model, layout)
+        free_params = slm.get_free_parameters(model)
+        torch_cache = slm.get_torch_cache(free_params, grad)
+        return slm, torch_cache
+
+    @staticmethod
+    def _check_copa_layout(layout: CircuitOutcomeProbabilityArrayLayout):
+        # I need to verify some assumptions on what layout.iter_unique_circuits()
+        # returns. Looking at the implementation of that function, the assumptions
+        # can be framed in terms of the "layout._element_indicies" OrderedDict.
+        eind = layout._element_indices
+        assert isinstance(eind, OrderedDict)
+        items = iter(eind.items())
+        k_prev, v_prev = next(items)
+        assert k_prev == 0
+        assert v_prev.start == 0
+        for k, v in items:
+            assert k == k_prev + 1
+            assert v.start == v_prev.stop
+            k_prev = k
+            v_prev = v
+        return v_prev.stop
+
+    def _bulk_fill_probs(self, array_to_fill, layout, stripped_abstractions: Optional[tuple] = None):
+        if stripped_abstractions is None:
+            slm, torch_cache = TorchForwardSimulator.separate_state(self.model, layout)
+        else:
+            slm, torch_cache = stripped_abstractions
+
+        layout_len = TorchForwardSimulator._check_copa_layout(layout)
+        probs = slm.circuit_probs(torch_cache)
+        array_to_fill[:layout_len] = probs.cpu().detach().numpy().flatten()
+        pass
+
+    def _bulk_fill_dprobs(self, array_to_fill, layout, pr_array_to_fill):
+        slm = StatelessModel(self.model, layout)
```

Comment:
Tiny note up front: I've renamed the variable that you're referencing as``stripped_abstractions``. It's now called ``split_model``. Now to answer your question ...

For _bulk_fill_probs (no "d") I was able to add an optional argument since there's a codepath in _bulk_fill_dprobs (with "d") that requires calling _bulk_fill_probs. In order to make use of an extra argument here we'd need to override the functions in the ForwardSimulator base class that call this function. I've added a TODO for us to explore this later.